spec/index_spec.rb

```diff
@@ -372,4 +372,51 @@
       )
     end
   end
+
+  describe Daru::CategoricalIndex do
+    before do
+      @idx = Daru::CategoricalIndex.new(
+        [:a, :b, :a, :a, :c])
+      @idx2 = Daru::CategoricalIndex.new(
+        [0, 1, 0, 0, 2])
+    end
+
+    context "#[]" do
+      context "when the category is non-numeric" do
+        it "retrives all elements belonging to that category" do
+          expect(@idx1[:a]).to eq(Daru::Index.new(
+            [0, 2, 3]))
+        end
+
+        it "retrives all elements belonging to given array of categories" do
+          expect(@idx1[:a, :c]).to eq(Daru::Index.new(
+            [0, 2, 3, 4]))
+        end
+
+        it "returns the position given positional index" do
+          expect(@idx1[0]).to eq(0)
+        end
+
+        it "raises exception given wrong positional index" do
+          expect { @indx1[5] }.to raise_error
+        end
+
+        it "returns the positions give positional index" do
+          expect(@idx1[0, 1, 2]).to eq(Daru::Index.new(
+            [0, 1, 2]))
+        end
+
+        it "raises exception given wrong positional indexes" do
+          expect { @indx1[0, 1, 5] }.to raise_error
+        end
+      end
+
+      context "when the category is numeric" do
+        it "retrives all elements belonging to that category" do
+          # The idea is to use #cat when retrieving numeric categories
+          expect(@idx2.cat[0]).to eq([0, 2, 3])
+        end
```

Comment:
I'm confused. @v0dro Could you summarize for `Daru::Index#[]` what you said?

Currently this is how `Daru::Index` works:

```ruby
[27] pry(main)> a = Daru::Index.new [1, :a, :b]
=> #<Daru::Index:0x0000000327f350
 @keys=[1, :a, :b],
 @relation_hash={1=>0, :a=>1, :b=>2},
 @size=3>
[28] pry(main)> a[:a]
=> 1
[29] pry(main)> a[1]
=> 0
[30] pry(main)> a[0]
=> 0
[31] pry(main)> a[3]
IndexError: Specified index 3 does not exist
from /home/ubuntu/workspace/daru/lib/daru/index.rb:94:in `[]'
```
To summarize, if value an index value, then it returns it positional index, otherwise it checks if it is a valid positional index or not. In case it is a valid positional index, it returns the value as it is.